seqr/views/apis/auth_api.py

```diff
@@ -35,6 +42,111 @@ def login_view(request):
 
     return create_json_response({'success': True})
 
+@csrf_exempt
+def login_google(request):
+  # Create flow instance to manage the OAuth 2.0 Authorization Grant Flow steps.
+  flow = google_auth_oauthlib.flow.Flow.from_client_config(
+      GOOGLE_AUTH_CLIENT_CONFIG, scopes=scopes)
+
+  # The URI created here must exactly match one of the authorized redirect URIs
+  # for the OAuth 2.0 client, which you configured in the API Console. If this
+  # value doesn't match an authorized URI, you will get a 'redirect_uri_mismatch'
+  # error.
+  flow.redirect_uri = GOOGLE_AUTH_CLIENT_CONFIG['web']['redirect_uris'][0]
+
+  authorization_url, state = flow.authorization_url(
+      # Enable offline access so that you can refresh an access token without
+      # re-prompting the user for permission. Recommended for web server apps.
+      access_type='offline',
+      # Enable incremental authorization. Recommended as a best practice.
+      include_granted_scopes='true')
+
+  # Store the state so the callback can verify the auth server response.
+  request.session['state'] = state
+
+  # Return the url to be sent for the Google Auth server by the the front end
+  return create_json_response({'data': authorization_url})
+
+@csrf_exempt
+def login_oauth2callback(request):
+    """
+    Callback function will be called after the OAuth2 server passes the authentication
+
+    In the previous function 'login_google', a redirect url is generated with the information of Google client ID and
+    the front end will be redirect to the url to do authentication with Google auth server. Once the authentication has
+    passed, the result of the authentication will be sent to the web server through this callback function.
+    :param request: includes the authorized code
+    :return: status of authorization server and resource server (which is AnVIL server) responses
+    """
+
+    # Specify the state when creating the flow in the callback so that it can
+    # verified in the authorization server response.
+    state = request.session['state']
+
+    flow = google_auth_oauthlib.flow.Flow.from_client_config(
```

Comment:
this seems copy-pasted from above. I think this should be instantiated once instead of on every request, but if you disagree it should at least be a helper function

Reply:
The difference is whether includes the `state` in the parameter. The previous one is to create a flow. This one is to recover a flow that has been created in the previous step of the flow.